src/test/Shared/Services/Login.service.test.tsx

```diff
@@ -0,0 +1,442 @@
+ * SOFTWARE.
+ */
+
+import { AuthMethod, LoginService, SessionState } from '@app/Shared/Services/Login.service';
+import { TargetService } from '@app/Shared/Services/Target.service';
+import { AuthCredentials } from '@app/Shared/Services/AuthCredentials.service';
+import { ApiV2Response } from '@app/Shared/Services/Api.service';
+import { SettingsService } from '@app/Shared/Services/Settings.service';
+import { firstValueFrom, of, timeout } from 'rxjs';
+import { fromFetch } from 'rxjs/fetch';
+
+jest.mock('rxjs/fetch', () => {
+  return {
+    fromFetch: jest.fn((url: any, opts: any): any => {}),
+  };
+});
+
+jest.unmock('@app/Shared/Services/Login.service');
+
+describe('Login.service', () => {
+  const mockFromFetch = fromFetch as jest.Mock;
+  let svc: LoginService;
+
+  describe('setLoggedOut', () => {
+    let authCreds: AuthCredentials;
+    let targetSvc: TargetService;
+    let settingsSvc: SettingsService;
+    let saveLocation: Location;
+
+    beforeAll(() => {
+      // Redirection is unimplemented in JSDOM, and cannot by spied on
+      saveLocation = window.location;
+      const locationMock = {
+        ...saveLocation,
+        replace: jest.fn(),
+      } as Location;
+      Object.defineProperty(window, 'location', {
+        value: locationMock,
+        writable: true,
+      });
+    });
+
+    afterAll(() => {
+      // Restore the original location
+      Object.defineProperty(window, 'location', {
+        value: saveLocation,
+        writable: false,
+      });
+    });
+
+    beforeEach(() => {
+      authCreds = {} as AuthCredentials;
+      targetSvc = {} as TargetService;
+      settingsSvc = new SettingsService();
+      (settingsSvc.webSocketDebounceMs as jest.Mock).mockReturnValue(0);
+    });
+
+    afterEach(() => {
+      sessionStorage.clear();
+    });
+
+    describe('with Basic AuthMethod', () => {
+      beforeEach(async () => {
+        const initAuthResp = createResponse(401, false, new Headers({ 'X-WWW-Authenticate': 'Basic' }), {
+          meta: {
+            type: 'text/plain',
+            status: 'Unauthorized',
+          },
+          data: {
+            reason: 'HTTP Authorization Failure',
+          },
+        });
+        const authResp = createResponse(200, true, new Headers({ 'X-WWW-Authenticate': 'Basic' }), {
+          meta: {
+            type: 'application/json',
+            status: 'OK',
+          },
+          data: {
+            result: {
+              username: 'user',
+            },
+          },
+        });
+        const logoutResp = createResponse(200, true);
+        mockFromFetch
+          .mockReturnValueOnce(of(initAuthResp))
+          .mockReturnValueOnce(of(authResp))
+          .mockReturnValueOnce(of(logoutResp));
+        const token = 'user:d74ff0ee8da3b9806b18c877dbf29bbde50b5bd8e4dad7a3a725000feb82e8f1';
+        window.location.href = 'https://example.com/';
+        location.href = window.location.href;
```

Comment:
Wondering if we need to set `location.ref`? Is it for completeness when setting `window.location.href`?

Reply:
`getTokenFromUrlFragment` and `hasBearerTokenUrlHash` use `location.hash` instead of `window.location.hash`. I wasn't sure if there was a meaningful difference or if they could be changed to `window.location`. Thoughts?

Reply:
I think `location` is just another shorthand reference to `window.location.hash`. I tend to use `window.*` for clarity :))